base/rw_mutex.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include "base/assert.hpp"
+
+#include "std/mutex.hpp"
+
+namespace threads
+{
+class RWMutex
+{
+public:
+  RWMutex() : m_counter(0) {}
+  void LockRead()
+  {
+    std::lock_guard<mutex> lockWriter(m_noWritersMutex);
+    {
+      std::lock_guard<mutex> lockCounter(m_counterMutex);
+      size_t const prevCounter = m_counter;
+      m_counter++;
+      if (prevCounter == 0)
```

Comment:
Is it possible to implement RWMutex via semaphores, like Preshing does? (https://github.com/preshing/cpp11-on-multicore/tree/master/common, http://preshing.com/20150316/semaphores-are-surprisingly-versatile/)